src/coordinator/coordinator_server.cc

```diff
@@ -0,0 +1,441 @@
+#include "src/coordinator/coordinator_server.h"
+
+#include <future>
+#include <sstream>
+#include <string>
+
+#include "absl/status/statusor.h"
+#include "absl/strings/str_cat.h"
+#include "grpc/grpc.h"
+#include "grpcpp/create_channel.h"
+#include "grpcpp/security/credentials.h"
+#include "grpcpp/server_context.h"
+#include "openssl/sha.h"
+#include "src/blockchain/two_phase_commit.h"
+#include "src/proto/cohort.grpc.pb.h"
+#include "src/proto/common.pb.h"
+#include "src/proto/coordinator.grpc.pb.h"
+#include "src/utils/grpc_utils.h"
+
+namespace coordinator {
+
+namespace {
+
+using Blockchain = ::blockchain::TwoPhaseCommit;
+using CohortStub = ::cohort::Cohort::StubInterface;
+using ::common::Namespace;
+using ::grpc::ClientContext;
+using ::grpc::ServerContext;
+
+enum TransactionType {
+  // Not an actual type.
+  UNSET_TRANSACTION,
+  READ_ONLY_TRANSACTION,
+  WRITE_ONLY_TRANSACTION,
```

Comment:
nit: I guess you can abandon `WRITE_ONLY_TRANSACTION` and only keep `READ_WRITE_TRANSACTION` since both of them require RW lock